atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketCloseTimestampCalculator.java

```diff
@@ -0,0 +1,148 @@
+    //
+    // We’re explicitly not doing this algorithm now given the added complexity, but this may be implemented if the
+    // fixed parameters are too coarse.
+    public CloseableState getCloseableStateForStartTimestamp(long startTimestamp) {
+        long openWallClockTimeMillis = puncherStore.getMillisForTimestamp(startTimestamp);
+        long closeWallClockTimeMillis = openWallClockTimeMillis + TIME_GAP_BETWEEN_BUCKET_START_AND_END.toMillis();
+
+        if (Instant.now(clock).toEpochMilli() < closeWallClockTimeMillis) {
+            return CloseableState.notCloseable();
+        }
+
+        long finalLogicalTimestamp = puncherStore.get(closeWallClockTimeMillis);
+        if (finalLogicalTimestamp <= startTimestamp) {
+            long freshTimestamp = freshTimestampSupplier.get();
+            if (freshTimestamp - startTimestamp < MIN_BUCKET_SIZE_FOR_NON_PUNCHER_CLOSE) {
+                log.info(
+                        NON_PUNCHER_CLOSE_LOG
+                                + " We read a fresh timestamp {}, but this is not sufficiently far from the start"
+                                + " timestamp to close the bucket.",
+                        SafeArg.of("finalTimestampFromPunchTable", finalLogicalTimestamp),
+                        SafeArg.of("startTimestamp", startTimestamp),
+                        SafeArg.of("timeGap", TIME_GAP_BETWEEN_BUCKET_START_AND_END),
+                        SafeArg.of("openWallClockTimeMillis", openWallClockTimeMillis),
+                        SafeArg.of("freshTimestamp", freshTimestamp));
+                return CloseableState.notCloseable();
+            }
+
+            long cappedTimestamp = Math.min(startTimestamp + MAX_BUCKET_SIZE_FOR_NON_PUNCHER_CLOSE, freshTimestamp);
+            if (cappedTimestamp != freshTimestamp) {
+                log.info(
+                        NON_PUNCHER_CLOSE_LOG
+                                + " We read a fresh timestamp {}, but this is too far from the start timestamp."
+                                + " Proposing a capped timestamp {} instead.",
+                        SafeArg.of("finalTimestampFromPunchTable", finalLogicalTimestamp),
+                        SafeArg.of("startTimestamp", startTimestamp),
+                        SafeArg.of("timeGap", TIME_GAP_BETWEEN_BUCKET_START_AND_END),
+                        SafeArg.of("openWallClockTimeMillis", openWallClockTimeMillis),
+                        SafeArg.of("freshTimestamp", freshTimestamp),
+                        SafeArg.of("cappedTimestamp", cappedTimestamp));
+            } else {
+                log.info(
+                        NON_PUNCHER_CLOSE_LOG
+                                + " We read a fresh timestamp {}, and this will be the proposed closing timestamp",
+                        SafeArg.of("finalTimestampFromPunchTable", finalLogicalTimestamp),
+                        SafeArg.of("startTimestamp", startTimestamp),
```

Comment:
TODO post self review - likely extract this into a separate method that does the logging so that we don't have to deal with arg ordering.